runtime/old_parser/type.go

```diff
@@ -0,0 +1,988 @@
+					if !ok || integerExpression.Value.Sign() < 0 {
+						p.reportSyntaxError("expected positive integer size for constant sized type")
+					} else {
+						size = integerExpression
+					}
+				}
+			}
+
+			p.skipSpaceAndComments()
+
+			endToken, err := p.mustOne(lexer.TokenBracketClose)
+			if err != nil {
+				return nil, err
+			}
+
+			typeRange := ast.NewRange(
+				p.memoryGauge,
+				startToken.StartPos,
+				endToken.EndPos,
+			)
+
+			if size != nil {
+				return ast.NewConstantSizedType(
+					p.memoryGauge,
+					elementType,
+					size,
+					typeRange,
+				), nil
+			} else {
+				return ast.NewVariableSizedType(
+					p.memoryGauge,
+					elementType,
+					typeRange,
+				), nil
+			}
+		},
+	)
+}
+
+func defineOptionalType() {
+	defineType(postfixType{
+		tokenType:    lexer.TokenQuestionMark,
+		bindingPower: typeLeftBindingPowerOptional,
+		leftDenotation: func(p *parser, left ast.Type, tokenRange ast.Range) ast.Type {
+			return ast.NewOptionalType(
+				p.memoryGauge,
+				left,
+				tokenRange.EndPos,
+			)
+		},
+	})
+
+	defineType(postfixType{
+		tokenType:    lexer.TokenDoubleQuestionMark,
+		bindingPower: typeLeftBindingPowerOptional,
+		leftDenotation: func(p *parser, left ast.Type, tokenRange ast.Range) ast.Type {
+			return ast.NewOptionalType(
+				p.memoryGauge,
+				ast.NewOptionalType(
+					p.memoryGauge,
+					left,
+					tokenRange.StartPos,
+				),
+				tokenRange.EndPos,
+			)
+		},
+	})
+}
+
+func defineReferenceType() {
+	defineType(prefixType{
+		tokenType:    lexer.TokenAmpersand,
+		bindingPower: typeLeftBindingPowerReference,
+		nullDenotation: func(p *parser, right ast.Type, tokenRange ast.Range) ast.Type {
+			return ast.NewReferenceType(
+				p.memoryGauge,
+				nil,
+				right,
+				tokenRange.StartPos,
+			)
+		},
+	})
+}
+
+func defineRestrictedOrDictionaryType() {
+
+	// For the null denotation it is not clear after the start
+	// if it is a restricted type or a dictionary type.
+	//
+	// If a colon is seen it is a dictionary type.
+	// If no colon is seen it is a restricted type.
+
+	setTypeNullDenotation(
+		lexer.TokenBraceOpen,
+		func(p *parser, startToken lexer.Token) (ast.Type, error) {
+
+			var endPos ast.Position
+
+			var dictionaryType *ast.DictionaryType
+			var restrictedType *ast.IntersectionType
+
+			var firstType ast.Type
+
+			atEnd := false
+
+			expectType := true
+
+			for !atEnd {
+				p.skipSpaceAndComments()
+
+				switch p.current.Type {
+				case lexer.TokenComma:
+					if dictionaryType != nil {
+						return nil, p.syntaxError("unexpected comma in dictionary type")
+					}
+					if expectType {
+						return nil, p.syntaxError("unexpected comma in restricted type")
+					}
+					if restrictedType == nil {
+						firstNominalType, ok := firstType.(*ast.NominalType)
+						if !ok {
+							return nil, p.syntaxError("non-nominal type in restriction list: %s", firstType)
+						}
+						restrictedType = ast.NewIntersectionType(
```

Comment:
Restricted type parsing in the old parser just ignores the `T` in `T{I}`. I think this is okay, but let me know if you think we should do this differently.

Reply:
We may need to have the `T` as well. Because `T{I}` and `R{I}` are considered two different types, and certain update checking rules, such as changing fields types, would rely on `T`

Reply:
I added a legacy field to `IntersectionType` in 6c6947ca7b24591700760b6aef8723a3822db0da to track this.